File: ios/Core/Domain/Sources/CoreDomain/MusicLibrary.swift
import Foundation

public protocol MusicProvider: Sendable {
    var id: MusicProviderId { get }

    func observeAvailability() -> AsyncStream<MusicServiceAvailability>
    func observePlaylists() -> AsyncStream<[PlaylistSummary]>
    func observePlaylist(id: PlaylistId) -> AsyncStream<Playlist?>
    func observeSavedTracks() -> AsyncStream<SavedTracksResult>

    func refreshAvailability() async throws
    func refreshPlaylists() async throws
    func refreshPlaylist(id: PlaylistId) async throws
    func refreshSavedTracks() async throws
}

public protocol MusicLibrary: Sendable {
    func observeAvailability() -> AsyncStream<MusicServiceAvailability>
    func observeAllPlaylists() -> AsyncStream<[PlaylistSummary]>
    func observePlaylist(ref: PlaylistRef) -> AsyncStream<Playlist?>
    func observeSavedTracks() -> AsyncStream<SavedTracksResult>

    func refreshAll() async throws
    func refreshPlaylist(ref: PlaylistRef) async throws
    func refreshSavedTracks() async throws
}

File: android/core/domain/src/main/kotlin/com/mplayeraudio/core/domain/musiclibrary/MusicLibrary.kt
package com.mplayeraudio.core.domain.musiclibrary

import kotlinx.coroutines.flow.Flow

interface MusicLibrary {
    fun observeAvailability(): Flow<MusicServiceAvailability>

    fun observeAllPlaylists(): Flow<List<PlaylistSummary>>

    fun observePlaylist(ref: PlaylistRef): Flow<Playlist?>

    fun observeSavedTracks(): Flow<SavedTracksResult>

    suspend fun refreshAll()

    suspend fun refreshPlaylist(ref: PlaylistRef)

    suspend fun refreshSavedTracks()
}
